In [1]:
###CONDUCT DATA ACCESS AND PROFILING###
##Importing and read doc
import pandas as pd
file_path = "/Users/haoranzhu/Downloads/Passengers_Satisfaction_Survey (2).csv"
df = pd.read_csv(file_path)
df.head()

/var/folders/33/ldfgcqdd5mn3kmrph8j1lm240000gn/T/ipykernel_70578/3196082834.py:5: DtypeWarning: Columns (14) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path)


,Gender,Customer Type,Age,Type of Travel,Class,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Food and drink,Inflight entertainment,Baggage handling,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes,Satisfaction
0,Male,Loyal Customer,69.0,Personal Travel,Business,32.0,3.0,2.0,3,5,1.0,1,3.0,14.0,0.0,Dissatisfied
1,Male,Disloyal Customer,20.0,Business travel,Eco,0.0,3.0,3.0,3,2,10.0,4,2.0,0.0,0.0,dissatisfied
2,Male,Disloyal Customer,41.0,Business travel,Business,0.0,1.0,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,neutral
3,Male,Loyal Customer,52.0,Business travel,Business,0.0,2.0,4.0,4,1,2.0,2,4.0,0.0,6.0,Dissatisfied
4,Female,Disloyal Customer,21.0,Business travel,Business,0.0,4.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,neutral


In [2]:
##Understand shape of the overall data 
df.shape

(104910, 16)

In [3]:
#Subset Data Frame Business Class
df_Business = df[df['Class'] == 'Business'].copy()
##Display First Rows
df_Business.head()

,Gender,Customer Type,Age,Type of Travel,Class,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Food and drink,Inflight entertainment,Baggage handling,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes,Satisfaction
0,Male,Loyal Customer,69.0,Personal Travel,Business,32.0,3.0,2.0,3,5,1.0,1,3.0,14.0,0.0,Dissatisfied
2,Male,Disloyal Customer,41.0,Business travel,Business,0.0,1.0,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,neutral
3,Male,Loyal Customer,52.0,Business travel,Business,0.0,2.0,4.0,4,1,2.0,2,4.0,0.0,6.0,Dissatisfied
4,Female,Disloyal Customer,21.0,Business travel,Business,0.0,4.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,neutral
5,Female,Loyal Customer,57.0,Business travel,Business,0.0,3.0,4.0,4,5,2.0,3,2.0,0.0,0.0,Dissatisfied


In [4]:
##Display Rows and Columns
df_Business.shape

(50000, 16)

In [5]:
##Columns Names and Data Types 
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
Index: 50000 entries, 0 to 104909
Data columns (total 16 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Gender                             48239 non-null  object 
 1   Customer Type                      48239 non-null  object 
 2   Age                                47373 non-null  float64
 3   Type of Travel                     48239 non-null  object 
 4   Class                              50000 non-null  object 
 5   Flight Distance                    46604 non-null  float64
 6   Inflight wifi service              48239 non-null  float64
 7   Departure/Arrival time convenient  48239 non-null  float64
 8   Ease of Online booking             42705 non-null  object 
 9   Food and drink                     42705 non-null  object 
 10  Inflight entertainment             43649 non-null  float64
 11  Baggage handling                   43649 non-null  object 

In [6]:
##Unique Values (Loop)
cols=df_Business.columns

for col in cols:
    print(col)
    unique=df_Business[col].unique()
    print(unique,'\n*\n\n')
    



Gender
['Male' 'Female' 'F' 'M' nan 'Fe male'] 
*


Customer Type
['Loyal Customer' 'Disloyal Customer' nan 'disloyal' 'Loyal'] 
*


Age
[ 69.  41.  52.  21.  57.  45.  39.  33.  10.  55.  59.  53.  56.  47.
  54.  25.  51.  60.  27.  43.  20.  19.  63.  50.  nan  38.  46.  85.
  26.  42.  30.  24.  22.  17.  48.  40.  16.  32.  35.  44.  49.  65.
  67.  66.  64.   7.  37.  36.  15.  62.  34.   0.  68.  70.  58.  11.
  74.  13.  80.  75.   9.  61.  71.  12.  73.  72.  18.  79.   8.  76.
  14.  77.  78. 180. 185.] 
*


Type of Travel
['Personal Travel' 'Business travel' 'Business' nan] 
*


Class
['Business'] 
*


Flight Distance
[3.200e+01 0.000e+00 5.600e+01 ... 4.983e+03 2.820e+06 9.000e+06] 
*


Inflight wifi service
[ 3.  1.  2.  4.  0.  5. nan] 
*


Departure/Arrival time convenient
[ 2.  1.  4.  0.  5.  3. nan] 
*


Ease of Online booking
['3' nan '4' '0' '5' 'Unknown' '2' '1'] 
*


Food and drink
['5' nan '1' '3' 'Great' '4' '2' '0'] 
*


Inflight entertainment
[ 1. nan  2.  3. 

In [7]:
##Missing Values
df_Business.isna().sum()

Gender                               1761
Customer Type                        1761
Age                                  2627
Type of Travel                       1761
Class                                   0
Flight Distance                      3396
Inflight wifi service                1761
Departure/Arrival time convenient    1761
Ease of Online booking               7295
Food and drink                       7295
Inflight entertainment               6351
Baggage handling                     6351
Cleanliness                          6351
Departure Delay in Minutes           6351
Arrival Delay in Minutes             7902
Satisfaction                         2160
dtype: int64

In [8]:
##SummaryStatistics
df_Business.describe(include='all')

,Gender,Customer Type,Age,Type of Travel,Class,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Food and drink,Inflight entertainment,Baggage handling,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes,Satisfaction
count,48239,48239,47373.000000,48239,50000,4.660400e+04,48239.000000,48239.000000,42705,42705,43649.000000,43649,43649.000000,43649.000000,42098.0,47840
unique,5,4,NaN,3,1,NaN,NaN,NaN,7,7,NaN,6,NaN,NaN,608.0,6
top,Female,Loyal Customer,NaN,Business travel,Business,NaN,NaN,NaN,3,4,NaN,4,NaN,NaN,0.0,satisfied
freq,23000,40770,NaN,44063,50000,NaN,NaN,NaN,8635,10890,NaN,17056,NaN,NaN,12777.0,30398
mean,NaN,NaN,41.853186,NaN,NaN,1.934729e+03,2.774519,2.910695,NaN,NaN,3.835391,NaN,3.737520,13.595363,NaN,NaN
std,NaN,NaN,13.730683,NaN,NaN,4.369344e+04,1.423917,1.501711,NaN,NaN,1.233508,NaN,0.987664,36.595048,NaN,NaN
min,NaN,NaN,0.000000,NaN,NaN,-5.421000e+03,0.000000,0.000000,NaN,NaN,0.000000,NaN,0.000000,-2.000000,NaN,NaN
25%,NaN,NaN,34.000000,NaN,NaN,5.940000e+02,2.000000,2.000000,NaN,NaN,3.000000,NaN,3.000000,0.000000,NaN,NaN
50%,NaN,NaN,43.000000,NaN,NaN,1.597000e+03,3.000000,3.000000,NaN,NaN,4.000000,NaN,4.000000,0.000000,NaN,NaN
75%,NaN,NaN,52.000000,NaN,NaN,2.565000e+03,4.000000,4.000000,NaN,NaN,5.000000,NaN,5.000000,10.000000,NaN,NaN


In [9]:
##Frequency Distribution
for col in cols:
    print(f"\nColumn: {col}")
    print(df_Business[col].value_counts(dropna=False))
    print("-" * 40)
    


Column: Gender
Gender
Female     23000
Male       22830
NaN         1761
F           1361
M           1047
Fe male        1
Name: count, dtype: int64
----------------------------------------

Column: Customer Type
Customer Type
Loyal Customer       40770
Disloyal Customer     6636
NaN                   1761
disloyal               831
Loyal                    2
Name: count, dtype: int64
----------------------------------------

Column: Age
Age
NaN      2627
39.0     1935
40.0     1677
41.0     1533
42.0     1514
         ... 
73.0       22
78.0       17
85.0       14
180.0       1
185.0       1
Name: count, Length: 75, dtype: int64
----------------------------------------

Column: Type of Travel
Type of Travel
Business travel    44063
Business            2122
Personal Travel     2054
NaN                 1761
Name: count, dtype: int64
----------------------------------------

Column: Class
Class
Business    50000
Name: count, dtype: int64
----------------------------------------

Column

In [10]:
##Duplicated Data Rows
df_Business.duplicated().sum()

np.int64(3064)

## Annotations of the Issues Data Presented ##

### **Data Issues Identified Before Cleansing** ###

**1. Inconsistent Column Names**  
Some columns include spaces and capital letters (e.g., “Customer Type”), which can cause errors in referencing column names during analysis.

**2. Inconsistent Categorical Values**  
- *Gender:* Includes inconsistent entries such as “F”, “M”, and “Fe male”.  
- *Customer Type:* Variants like “Loyal Customer”, “Disloyal Customer”, and “disloyal” need to be standardized.  
- *Type of Travel:* Contains “Business travel”, “Personal Travel”, and “Business”.

**3. Outliers and Invalid Numeric Data**  
- *Age:* Includes unrealistic values like 7 and 185.  
- *Flight Distance:* Contains extreme outliers (e.g., 2,820,000 and 9,000,000).

**4. Improper Rating Scales**  
Rating columns (Inflight WiFi Service, Cleanliness, etc.) contain text entries like “Great” and invalid values like 0 or 10.  
The valid range should be **1–5**.

**5. Mixed Data Types**  
Some numeric columns are stored as objects due to the presence of text or symbols.

**6. Missing Values (NaN)**  
Several columns, including *Ease of Online Booking*, *Food and Drink*, *Inflight Entertainment*, *Baggage Handling*, *Cleanliness*, and *Age*, have missing values.

**7. Duplicated Data**  
The dataset contains 3,864 duplicated rows, which must be removed to prevent bias.

**8. Ambiguous Satisfaction Labels**  
The *Satisfaction* column contains inconsistent values such as “satisfied”, “Satisfied”, “neutral”, and numeric codes like 0 and 1, requiring standardization.


In [11]:
###CONDUCT DATA CLEANSING###
##Verify df_Business Columns
for col in df_Business.columns:
    print(repr(col))

'Gender'
'Customer Type'
'Age'
'Type of Travel'
'Class'
'Flight Distance'
'Inflight wifi service'
'Departure/Arrival time convenient'
'Ease of Online booking'
'Food and drink'
'Inflight entertainment'
'Baggage handling'
'Cleanliness'
'Departure Delay in Minutes'
'Arrival Delay in Minutes'
'Satisfaction'


In [12]:
##Verify columns from the original Database
for c in df.columns: print(repr(c))

'Gender'
'Customer Type'
'Age'
'Type of Travel'
'Class'
'Flight Distance'
'Inflight wifi service'
'Departure/Arrival time convenient'
'Ease of Online booking'
'Food and drink'
'Inflight entertainment'
'Baggage handling'
'Cleanliness'
'Departure Delay in Minutes'
'Arrival Delay in Minutes'
'Satisfaction'


In [13]:
##Drop 'Class' Column
df_Business = df_Business.drop(columns=['Class'], errors='ignore')

In [14]:
#Verify column was dropped
df_Business.columns

Index(['Gender', 'Customer Type', 'Age', 'Type of Travel', 'Flight Distance',
       'Inflight wifi service', 'Departure/Arrival time convenient',
       'Ease of Online booking', 'Food and drink', 'Inflight entertainment',
       'Baggage handling', 'Cleanliness', 'Departure Delay in Minutes',
       'Arrival Delay in Minutes', 'Satisfaction'],
      dtype='object')

In [15]:
##Rename Columns in Lowercase and Remove Space and Replace it with _
df_Business.rename(columns = lambda x: x.lower().replace(' ','_'), inplace= True)
df_Business.columns

Index(['gender', 'customer_type', 'age', 'type_of_travel', 'flight_distance',
       'inflight_wifi_service', 'departure/arrival_time_convenient',
       'ease_of_online_booking', 'food_and_drink', 'inflight_entertainment',
       'baggage_handling', 'cleanliness', 'departure_delay_in_minutes',
       'arrival_delay_in_minutes', 'satisfaction'],
      dtype='object')

In [16]:
##Modify Improperly Formatted Data and Handle Bad Data
#Column 1 'Gender' - Standardized abbreviations ("F"/"M") and spacing issues ("Fe male") to "Female"/"Male"
df_Business['gender'] = df_Business['gender'].astype(str).str.strip()
df_Business['gender'] = df_Business['gender'].replace({
    'F':'Female','M':'Male','Fe male':'Female'})



In [17]:
df_Business['gender'].unique()

array(['Male', 'Female', 'nan'], dtype=object)

In [18]:
#Column 2 'customer_type' - Unified similar categories ("Loyal Customer"/"Disloyal Customer"/"disloyal") into "Loyal"/"Disloyal"
df_Business['customer_type'] = df_Business['customer_type'].replace({
    'Loyal Customer':'Loyal',
    'Disloyal Customer':'Disloyal',
    'disloyal':'Disloyal'})


In [19]:
df_Business['customer_type'].unique()

array(['Loyal', 'Disloyal', nan], dtype=object)

In [20]:
import numpy as np

In [21]:
#Column 3 'age' - Ensuring age fits a reasonable range 
df_Business.loc[~df_Business['age'].between(12, 90), 'age'] = np.nan

In [22]:
df_Business['age'].unique()

array([69., 41., 52., 21., 57., 45., 39., 33., nan, 55., 59., 53., 56.,
       47., 54., 25., 51., 60., 27., 43., 20., 19., 63., 50., 38., 46.,
       85., 26., 42., 30., 24., 22., 17., 48., 40., 16., 32., 35., 44.,
       49., 65., 67., 66., 64., 37., 36., 15., 62., 34., 68., 70., 58.,
       74., 13., 80., 75., 61., 71., 12., 73., 72., 18., 79., 76., 14.,
       77., 78.])

In [23]:
#Column 4 'type_of_travel' - Standardized travel categories ("Business"-"Business Travel") for naming consistency 
df_Business['type_of_travel'] = df_Business['type_of_travel'].replace({
    'Business':'Business travel',
    'Personal Travel':'Personal Travel'})

In [24]:
df_Business['type_of_travel'].unique()

array(['Personal Travel', 'Business travel', nan], dtype=object)

In [25]:
#Column 5 'flight_distance' - Get rid of outiers in flight distance 
df_Business.loc[df_Business['flight_distance'] <= 0, 'flight_distance'] = np.nan
lo, hi = df_Business['flight_distance'].quantile([0.01, 0.99])
df_Business['flight_distance'] = df_Business['flight_distance'].clip(lo, hi)


In [26]:
df_Business['flight_distance'].unique()

array([ 116.,   nan,  118., ..., 3942., 3943., 3944.])

In [27]:
#Columns 6-13 - Rate ordinal values (0-5) 
rating_rest = [
    'inflight_wifi_service',
    'departure/arrival_time_convenient',
    'ease_of_online_booking',
    'food_and_drink',
    'inflight_entertainment',
    'baggage_handling',
    'cleanliness'
]

rating_rest = [c for c in rating_rest if c in df_Business.columns]

for c in rating_rest:
    df_Business[c] = pd.to_numeric(df_Business[c], errors='coerce')
    df_Business.loc[~df_Business[c].between(0, 5), c] = np.nan

In [28]:
#Columns 6-13 - Adjust rating scale from 0-5 to 1-5; converting 0 to NaN
rating_cols = [
    'inflight_wifi_service',
    'departure/arrival_time_convenient',
    'ease_of_online_booking',
    'food_and_drink',
    'inflight_entertainment',
    'baggage_handling',
    'cleanliness'
]


for c in rating_cols:
    df_Business[c] = pd.to_numeric(df_Business[c], errors='coerce')
    df_Business.loc[(df_Business[c] == 0) | (~df_Business[c].between(1, 5)), c] = np.nan

In [29]:
###Check the corect columns were cleaned   
rating_rest

['inflight_wifi_service',
 'departure/arrival_time_convenient',
 'ease_of_online_booking',
 'food_and_drink',
 'inflight_entertainment',
 'baggage_handling',
 'cleanliness']

In [30]:
###Check each cleaned columns unique values; to make sure no out-of-range values
for c in rating_rest:
    print(f"\n{c} unique values:")
    print(sorted(df_Business[c].dropna().unique()))



inflight_wifi_service unique values:
[np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0), np.float64(5.0)]

departure/arrival_time_convenient unique values:
[np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0), np.float64(5.0)]

ease_of_online_booking unique values:
[np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0), np.float64(5.0)]

food_and_drink unique values:
[np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0), np.float64(5.0)]

inflight_entertainment unique values:
[np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0), np.float64(5.0)]

baggage_handling unique values:
[np.float64(1.0), np.float64(2.0), np.float64(3.0), np.float64(4.0), np.float64(5.0)]

cleanliness unique values:
[np.float64(2.0), np.float64(3.0), np.float64(4.0), np.float64(5.0)]


In [31]:
###Check how many missing values were created after cleaning
df_Business[rating_rest].isna().sum()

inflight_wifi_service                3650
departure/arrival_time_convenient    3885
ease_of_online_booking               9539
food_and_drink                       7815
inflight_entertainment               6671
baggage_handling                     7025
cleanliness                          6355
dtype: int64

In [32]:
###Confirm numeric Integrity - min (1) max (5); Show a descriptive summary of the cleaned columns
df_Business[rating_rest].describe()

,inflight_wifi_service,departure/arrival_time_convenient,ease_of_online_booking,food_and_drink,inflight_entertainment,baggage_handling,cleanliness
count,46350.000000,46115.000000,40461.000000,42185.000000,43329.000000,42975.000000,43645.000000
mean,2.887594,3.044758,3.037691,3.437857,3.791248,3.888493,3.737862
std,1.335533,1.396714,1.370849,1.196997,1.119549,1.101006,0.987061
min,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,2.000000
25%,2.000000,2.000000,2.000000,2.000000,3.000000,3.000000,3.000000
50%,3.000000,3.000000,3.000000,4.000000,4.000000,4.000000,4.000000
75%,4.000000,4.000000,4.000000,4.000000,5.000000,5.000000,5.000000
max,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000


In [33]:
#Column 14 'departure_delay_in_minutes'- Convert into numbers, convert value less then zero into NaN, get rid of extreme value
for c in ['departure_delay_in_minutes', 'arrival_delay_in_minutes']:
    df_Business[c] = pd.to_numeric(df_Business[c], errors='coerce')
    df_Business.loc[df_Business[c] < 0, c] = np.nan
    lo, hi = df_Business[c].quantile([0.01, 0.99])
    df_Business[c] = df_Business[c].clip(lo, hi)



In [34]:
###Verify result 
df_Business[['departure_delay_in_minutes', 'arrival_delay_in_minutes']].describe()

,departure_delay_in_minutes,arrival_delay_in_minutes
count,43152.000000,41422.000000
mean,12.899866,13.114263
std,29.186623,30.093039
min,0.000000,0.000000
25%,0.000000,0.000000
50%,0.000000,0.000000
75%,11.000000,11.000000
max,172.000000,177.000000


In [35]:
#Column 15 'Satisfaction' - Reconstruct the label 
df_Business['satisfaction'] = (
        df_Business['satisfaction']
          .replace({'satisfied':'Satisfied',
                    'dissatisfied':'Dissatisfied',
                    'neutral':'Neutral',
                    '1':'Satisfied',
                    '0':'Dissatisfied'}))

In [36]:
##Verify the Output
df_Business['satisfaction'].unique()

array(['Dissatisfied', 'Neutral', 'Satisfied', nan], dtype=object)

In [37]:
##Handle Missing Values
#Categorical Values - Fill with 'Unknown'
for col in ['customer_type', 'type_of_travel', 'gender']:
    if col in df_Business.columns:
        df_Business[col] = df_Business[col].fillna('Unknown')


In [38]:
#Continuos Numeric Values - Fill with median
for col in ['age', 'flight_distance']:
    if col in df_Business.columns:
        df_Business[col] = df_Business[col].fillna(df_Business[col].median())

In [39]:
#Rating Columns (0-5) - Fill with mode
rating_cols = [
    'inflight_wifi_service',
    'departure/arrival_time_convenient',
    'ease_of_online_booking',
    'food_and_drink',
    'inflight_entertainment',
    'baggage_handling',
    'cleanliness'
]
rating_cols = [c for c in rating_cols if c in df_Business.columns]

for col in rating_cols:
    mode_val = df_Business[col].mode(dropna=True).iat[0]
    df_Business[col] = df_Business[col].fillna(mode_val)

In [40]:
#Delay Columns - Fill with median 
for col in ['departure_delay_in_minutes', 'arrival_delay_in_minutes']:
    if col in df_Business.columns:
        df_Business[col] = df_Business[col].fillna(df_Business[col].median())

In [41]:
#Satisfaction - Fill with "Unknown"
if 'satisfaction' in df_Business.columns:
    df_Business['satisfaction'] = df_Business['satisfaction'].fillna('Unknown')

In [42]:
##Verify that no missing values remain 
print(df_Business.isna().sum().sort_values(ascending=False).head(10))
print("Total NA:", int(df_Business.isna().sum().sum()))

gender                               0
customer_type                        0
age                                  0
type_of_travel                       0
flight_distance                      0
inflight_wifi_service                0
departure/arrival_time_convenient    0
ease_of_online_booking               0
food_and_drink                       0
inflight_entertainment               0
dtype: int64
Total NA: 0


In [43]:
##Remove fully duplicated observations
df_Business = df_Business.drop_duplicates()

In [44]:
#Verify
print("Duplicates after drop:", df_Business.duplicated().sum())
print("Final row count:", len(df_Business))

Duplicates after drop: 0
Final row count: 46864


In [45]:
##Fix Column Datatypes
#Convert rating columns from object to float
cols_to_float = ['ease_of_online_booking', 'food_and_drink', 'baggage_handling']
for col in cols_to_float:
    if col in df_Business.columns:
        df_Business[col] = df_Business[col].astype(float)

In [46]:
#Convert categorical values to category dtype
cat_cols = ['gender', 'customer_type', 'type_of_travel', 'satisfaction']
for col in cat_cols:
    if col in df_Business.columns:
        df_Business[col] = df_Business[col].astype('category')

In [47]:
#Verify Datatypes
df_Business.dtypes

gender                               category
customer_type                        category
age                                   float64
type_of_travel                       category
flight_distance                       float64
inflight_wifi_service                 float64
departure/arrival_time_convenient     float64
ease_of_online_booking                float64
food_and_drink                        float64
inflight_entertainment                float64
baggage_handling                      float64
cleanliness                           float64
departure_delay_in_minutes            float64
arrival_delay_in_minutes              float64
satisfaction                         category
dtype: object

## Final Annotations — Comparing the Data Before and After Cleansing 

The original dataset went through multiple data cleansing processes to ensure **consistency, accuracy, and readiness for analysis**.

---

### **Summary of Column Changes**

**Gender**  
Originally stored as text (`object`) with inconsistent entries such as “Fe male”, “M”, and “F”. These were standardized to “Male” and “Female”, converted to categorical type, and all missing values were filled with **'Unknown'**.

**Customer_Type**  
Contained mixed formats (“Loyal Customer”, “Disloyal”, and lowercase variants). Categories were unified into **“Loyal”** and **“Disloyal”**, datatype converted to categorical, and missing values filled with **'Unknown'**.

**Age**  
Contained out-of-range values (7 and 185) and some missing entries. Values outside the realistic range (12–90) were replaced with NaN, then filled with the **median age**. Datatype remained numeric (`float`).

**Type_of_Travel**  
Included inconsistent categories (“Business”, “Business travel”) and missing data. These were standardized to **“Business travel”** and **“Personal Travel”**, converted to categorical, and missing values replaced with **'Unknown'**.

**Flight_Distance**  
Contained extreme outliers and missing values. Outliers were capped within the **1st and 99th percentiles**, datatype preserved as numeric (`float`), and missing entries filled with the **median**.

**Inflight_Wifi_Service**  
Originally numeric but included values outside the 1–5 range. Values were restricted to valid ratings (**1–5**), converted to numeric, and missing values filled with the **mode**.

**Departure/Arrival_Time_Convenient**  
Contained out-of-range and missing values. All values were limited to the **1–5 range**, datatype set to `float`, and missing entries replaced with the **mode**.

**Ease_of_Online_Booking**  
Read as object type due to text and “Unknown” entries. Converted to `float`, cleaned of non-numeric values, restricted to **1–5 range**, and filled with the **mode**.

**Food_and_Drink**  
Contained text entries like “Great” and numeric strings. Converted to numeric, restricted to the **1–5 range**, and missing values filled with the **mode**.

**Inflight_Entertainment**  
Included invalid ratings (e.g., 10) and missing values. Out-of-range values were replaced with NaN, restricted to **1–5**, datatype set to `float`, and missing entries filled with the **mode**.

**Baggage_Handling**  
Contained invalid entry “A” and some missing values. Converted to `float`, invalid entries replaced with NaN, restricted to **1–5**, and missing data filled with the **mode**.

**Cleanliness**  
Numeric but with some missing entries and minor inconsistencies. Restricted to the **1–5 range**, missing values filled with **mode**, and datatype kept as `float`.

**Departure_Delay_in_Minutes**  
Included negative and extreme delay times. Converted to numeric, negative values replaced with NaN, outliers capped within the **1st–99th percentile range**, and missing entries filled with the **median**.

**Arrival_Delay_in_Minutes**  
Similar to departure delay, negative and extreme values were cleaned, datatype converted to `float`, and missing entries replaced with the **median**.

**Satisfaction**  
Contained inconsistent labels (“1”, “0”, “satisfied”, “neutral”) and missing values. All categories standardized to **“Satisfied”**, **“Dissatisfied”**, and **“Neutral”**, datatype changed to categorical, and missing entries filled with the **most frequent label**.

**Class**  
This column was dropped as it was no longer needed for analysis.

---

### **Post-Cleansing Summary**

After cleansing, all columns are standardized to their correct datatypes, inconsistent entries were corrected, and **no missing values remain in the dataset**.  
The data is now **clean, consistent, and ready for analysis**.
**Overall Comparison Summary:**  
The data cleansing process successfully improved dataset quality by eliminating inconsistencies, correcting data types, and handling missing values. Compared to the original dataset, the cleaned version is fully standardized, reliable, and ready for analytical modeling.




In [48]:
df_Business.to_csv('cleaned_data.csv', index=False)

In [49]:
import os
os.getcwd()

'/Users/haoranzhu/Downloads'

In [50]:
os.listdir()

['overcoming-two-issues-that-are-sinking-gen-ai-programs_final-4.pdf',
 'I20_Zhu_Haoran_N0031423284_Post-completion OPT I-20 for USCIS Submission.pdf',
 'BrattleGroup_CoverLetter_LeoZhu.docx',
 'Resume_ Leo Zhu.docx',
 'Leo_Zhu_JPM_Cover_Letter.docx',
 'OperaSetup.zip',
 'MISM6202_05a_Probability_Fundamentals.pdf',
 'bain-report__midyear-private-equity-2025-4.pdf',
 '_______________.csv',
 'Resume_ Leo Zhu-3.pdf',
 'googlechrome.dmg',
 'CBRE_Investment_Internship_Cover_Letter.docx',
 'MS_in_Business_Analytics_Fact_Sheet-8.pdf',
 'the-evolution-of-commerce-media-navigating-a-new-era-in-advertising-33.pdf',
 'the-evolution-of-commerce-media-navigating-a-new-era-in-advertising-27.pdf',
 'the-evolution-of-commerce-media-navigating-a-new-era-in-advertising-4.pdf',
 'the-state-of-fashion-beauty-june-2025-6.pdf',
 '截屏2025-09-23 17.40.17.png',
 'Leo Zhu-resume.pdf',
 'Module 8 - Lab - Cross Validation.R',
 'the-state-of-fashion-beauty-june-2025-7.pdf',
 'Exam_1_Reviewing_Sheet.docx',
 'item-m4